src/util/pg.py

```diff
@@ -361,9 +361,18 @@ def alter_column_type(cr, table, column, type, using=None, logger=_logger):
             [model_of_table(cr, table), column],
         )
     if not using:
-        # Simple case. Use general SQL syntax
-        cr.execute('ALTER TABLE "%s" ALTER COLUMN "%s" TYPE %s' % (table, column, type))
-        return
+        # if there is a high number of NULL entries, it will be faster to just ignore those
+        cr.execute(
+            "SELECT null_fract FROM pg_stats WHERE schemaname = current_schema() AND tablename = %s AND attname = %s",
+            [table, column],
+        )
+        [null_frac] = cr.fetchone()
```

Comment:
`pg_stats` may return nothing.
```suggestion
        [null_frac] = cr.fetchone() or (0.0,)
```

Reply:
Only if the column does not exist. And in this case, that's the next query that will fail.

Reply:
Actually, it may be missing without calling `ANALYZE` before.

Reply:
The ANALYZE by itself can be expensive. Usually, we call ANALYZE on the whole DB very early in the upgrade and stats should already be there. How about only doing it here, if there are no stats for the column already?

Reply:
> Actually, it may be missing without calling ANALYZE before.

Yes, and even with a full ANALYZE at the beginning of the upgrade, newly created tables/columns are not reflected in pg_stats until next checks.

I think analyzing the column is a good approach. We are going to perform a potentially heavy operation on the column anyways but the question is: would this defeat the purpose of this optimization? @KangOl I guess you have some test run we can get more data from. If the analyze is as expensive as calling alter type then maybe a good compromise is analyzing only if there is nothing in pg_stats as @cawo-odoo suggested.

Reply:
> That's why I only `analyze`d this column.

I noticed that, my thought was just that in the vast majority of cases[^1] it should be superfluous to call it at all, such that whatever the cost, it is wasted. Of course you can argue that the cost is so small compared to the actual operation's, that we don't care. You decide.

[^1]: I know that internal disables the full ANALYZE sometimes, but in the platform, all columns that exist in the original DB should already have stats. And of course now that this repo is public, there are also users where we don't know about that.

Reply:
Internal don' t analyse because stats are accurate at the beginning of the upgrade. But it' s a specific use case